### PR TITLE
[SPARK-12569][PySpark][ML]:DecisionTreeRegressor: provide variance of prediction: Python API

### PR DESCRIPTION
**Pull request by wangmiao1981**
## What changes were proposed in this pull request?

A new column VarianceCol has been added to DecisionTreeRegressor in ML scala code.

This patch adds the corresponding Python API, HasVarianceCol, to class DecisionTreeRegressor.
## How was this patch tested?

./dev/lint-python 
PEP8 checks passed.
rm -rf _build/*
pydoc checks passed.

./python/run-tests --python-executables=python2.7 --modules=pyspark-ml
Running PySpark tests. Output is in /Users/mwang/spark_ws_0904/python/unit-tests.log
Will test against the following Python executables: ['python2.7']
Will test the following Python modules: ['pyspark-ml']
Finished test(python2.7): pyspark.ml.evaluation (12s)
Finished test(python2.7): pyspark.ml.clustering (18s)
Finished test(python2.7): pyspark.ml.classification (30s)
Finished test(python2.7): pyspark.ml.recommendation (28s)
Finished test(python2.7): pyspark.ml.feature (43s)
Finished test(python2.7): pyspark.ml.regression (31s)
Finished test(python2.7): pyspark.ml.tuning (19s)
Finished test(python2.7): pyspark.ml.tests (34s)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by holdenk** [comment]
Hi @wangmiao1981 thanks for starting this always good to see more people working to keep the Python API up to date :) For the params that live in `shared.py` they are autogenerated based on `_shared_params_code_gen.py` so please update that file and then generate the param rather than manually updating `shared.py`.

**Comment by wangmiao1981** [comment]
@holdenk Thanks for pointing it out. I will revise it.

**Comment by wangmiao1981** [comment]
@holdenk I made the changes and tested the gen code. Can you review it? Thanks!

File: python/pyspark/ml/param/_shared_params_code_gen.py [hunk 1]
```diff
@@ -146,7 +146,9 @@ def get$Name(self):
         ("weightCol", "weight column name. If this is not set or empty, we treat " +
          "all instance weights as 1.0.", None, "TypeConverters.toString"),
         ("solver", "the solver algorithm for optimization. If this is not set or empty, " +
-         "default value is 'auto'.", "'auto'", "TypeConverters.toString")]
+         "default value is 'auto'.", "'auto'", "TypeConverters.toString"),
+        ("varianceCol", "column name for the biased sample variance" +
```
Comment (holdenk) [inline python/pyspark/ml/param/_shared_params_code_gen.py]: I think you could merge the description to fit on the first line and do the linebreak after None - would make it a bit easier to read (but double check with the pylint script of course)

File: python/pyspark/ml/regression.py [hunk 2]
```diff
@@ -433,12 +433,12 @@ class DecisionTreeRegressor(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredi
     def __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction",
                  maxDepth=5, maxBins=32, minInstancesPerNode=1, minInfoGain=0.0,
                  maxMemoryInMB=256, cacheNodeIds=False, checkpointInterval=10, impurity="variance",
-                 seed=None):
+                 seed=None, variance="varianceCol"):
```
Comment (holdenk) [inline python/pyspark/ml/regression.py]: So I think the default here should probably be None to match `DecisionTreeRegressor.scala`.

**Comment by holdenk** [comment]
Two minor comments, we could also probably add a simple test to the existing decission tree tests in tests.py (nothing too complicated just make sure the value is respected - which might be overkill but having a little extra test logic makes it easier to catch changes in the Scala API we need to also update on the Python side).

**Comment by wangmiao1981** [comment]
@holdenk Thanks for your comments! I will make changes accordingly.

**Comment by wangmiao1981** [comment]
@jkbradley Can you add me to white list to trigger the integration test?

Thanks!

Miao

**Comment by holdenk** [comment]
@wangmiao1981 might still be good to add a simple test? But yes we should run this in jenkins if one of the commiters can whitelist.

**Comment by wangmiao1981** [comment]
@holdenk I am think what tests should be added. Do you have any suggestions?

Thanks!

Miao

**Comment by MLnick** [comment]
Jenkins add to whitelist

**Comment by jkbradley** [comment]
I'd recommend updating the docstring test to look at the variance column and print its value for one row.

File: python/pyspark/ml/regression.py [hunk 3]
```diff
@@ -425,6 +425,10 @@ class DecisionTreeRegressor(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredi
     True
     >>> model.depth == model2.depth
     True
+    >>> isinstance(dt.setVarianceCol("varianceColTest"), DecisionTreeRegressor)
+    True
+    >>> dt.getVarianceCol() == "varianceColTest"
+    True
```
Comment (MLnick) [inline python/pyspark/ml/regression.py]: This is really just testing set/get of params. I think what @jkbradley was intending is something like L414 above where `model.transform(test1).head().prediction` checks extracting the value for prediction col for one row. You should be able to do something similar here.

Comment (wangmiao1981) [inline python/pyspark/ml/regression.py]: Got it! Thanks! I will revise it.

File: python/pyspark/ml/regression.py [hunk 4]
```diff
@@ -433,12 +440,12 @@ class DecisionTreeRegressor(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredi
     def __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction",
                  maxDepth=5, maxBins=32, minInstancesPerNode=1, minInfoGain=0.0,
                  maxMemoryInMB=256, cacheNodeIds=False, checkpointInterval=10, impurity="variance",
-                 seed=None):
+                 seed=None, variance=None):
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: Should be ```varianceCol```.

File: python/pyspark/ml/regression.py [hunk 5]
```diff
@@ -433,12 +440,12 @@ class DecisionTreeRegressor(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredi
...
         """
         __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction", \
                  maxDepth=5, maxBins=32, minInstancesPerNode=1, minInfoGain=0.0, \
                  maxMemoryInMB=256, cacheNodeIds=False, checkpointInterval=10, \
-                 impurity="variance", seed=None)
+                 impurity="variance", seed=None, variance=None)
         """
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: ditto

File: python/pyspark/ml/regression.py [hunk 6]
```diff
@@ -454,12 +461,12 @@ def __init__(self, featuresCol="features", labelCol="label", predictionCol="pred
     def setParams(self, featuresCol="features", labelCol="label", predictionCol="prediction",
                   maxDepth=5, maxBins=32, minInstancesPerNode=1, minInfoGain=0.0,
                   maxMemoryInMB=256, cacheNodeIds=False, checkpointInterval=10,
-                  impurity="variance", seed=None):
+                  impurity="variance", seed=None, variance=None):
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: ditto

File: python/pyspark/ml/regression.py [hunk 7]
```diff
@@ -454,12 +461,12 @@ def __init__(self, featuresCol="features", labelCol="label", predictionCol="pred
...
         """
         setParams(self, featuresCol="features", labelCol="label", predictionCol="prediction", \
                   maxDepth=5, maxBins=32, minInstancesPerNode=1, minInfoGain=0.0, \
                   maxMemoryInMB=256, cacheNodeIds=False, checkpointInterval=10, \
-                  impurity="variance", seed=None)
+                  impurity="variance", seed=None, variance=None)
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: ditto

**Comment by yanboliang** [comment]
+1 @jkbradley

File: python/pyspark/ml/regression.py [hunk 8]
```diff
@@ -425,6 +429,9 @@ class DecisionTreeRegressor(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredi
     True
     >>> model.depth == model2.depth
     True
+    >>> test2 = sqlContext.createDataFrame([(Vectors.sparse(1, [1.0], [1.0]),)], ["features"])
+    >>> model.transform(test2).head().varianceColTest
+    0.0
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: We can test the output variance is correct by:
```
>>> model.setVarianceCol("variance")
>>> model.transform(test1).head().variance
```

File: python/pyspark/ml/regression.py (hunk 4 above)
Comment (wangmiao1981) [inline python/pyspark/ml/regression.py]: This column doesn't have a default value, as pointed by Holden.

**Comment by yanboliang** [comment]
LGTM cc @jkbradley @mengxr

**Comment by MLnick** [comment]
LGTM too

**Comment by holdenk** [comment]
+1/LGTM

**Comment by jkbradley** [comment]
LGTM
Merging with master
Thanks all for the PR and reviews!

**Pull request closed without merging**